added input order preservation and exclusion list to get_equations_for

MauriceHendrix opened the pull request:
I have had a play with get_equations_for and it's not straightforward to preserve an order resembling the input list due to how the digraph works. 
This way does seem seem to work, although it feels like it can be written up nicer maybe. Specifically it keeps also the original sorting as the default option (hence only acts differently if you provide a flag), not sure if that's useful.

Unfortunately `graph.predecessors(parent)` doesn't always return things in the same order, hence the sorting. 

Also: I'm assuming that if I get predecessors all those don't have any interdependence. I.e. one of the predecessors shouldn't contain another one, as then it should be a parent? If that's not necessarily the case, we would need a more complex sorting.

skeating commented on cellmlmanip/model.py:
```diff
@@ -410,27 +410,44 @@ def get_equation_graph(self, refresh=False) -> nx.DiGraph:
         return graph
 
-    def get_equations_for(self, symbols, lexicographical_sort=True):
+    def get_equations_for(self, symbols, sort_by_input_symbols=False, excluded_symbols=[]):
```
Normally I would say it was not ideal to change the function signature - but since it is entirely internal I think its okay.

MauriceHendrix replied:
Yes good idea feel free to change :)

skeating commented on tests/test_hodgkin.py:
```diff
@@ -276,34 +276,40 @@ def test_has_ontology_annotation(self, model):
+             membrane_stimulus_current_offset, membrane_stimulus_current_period],
+            True, [membrane_stimulus_current_period])
+
+        # There should be 10 in this model
```
10 or 5 ?

MauriceHendrix replied:
5 sorry

skeating commented on tests/test_hodgkin.py:
```diff
@@ -276,34 +276,40 @@ def test_has_ontology_annotation(self, model):
                            sympy.Dummy('sodium_channel_m_gate$m') ** 3.0 * sympy.Dummy('sodium_channel$g_Na') *
                            sympy.Dummy('sodium_channel_h_gate$h') * (sympy.Dummy('membrane$V') -
-                           sympy.Dummy('sodium_channel$E_Na')))
-                  ]
```
I would leave the close bracket on a separate line as it makes it much clearer when doing a diff what has actually changed

MauriceHendrix replied:
ok

skeating commented on tests/test_hodgkin.py:
```diff
@@ -276,34 +276,40 @@ def test_has_ontology_annotation(self, model):
+
+        equations = model.get_equations_for([membrane_fast_sodium_current, membrane_stimulus_current_offset])
+        input_ordered_equations = model.get_equations_for(
+            [membrane_fast_sodium_current, membrane_stimulus_current_offset,
```
should membrane_stimulus_current_offset be here twice

MauriceHendrix replied:
it's there twice to check that this doesn't lead to duplication of equations even if the input list perhaps contains duplicates.
not 100% sure it's necessary to be fair.

skeating replied:
If there is a possibility that duplication in the input list would cause duplication of equations; then this is really a separate test. I would do the call with no duplication as well and also check list with duplication produces the correct result

MauriceHendrix replied:
ok added.

skeating commented on cellmlmanip/model.py:
```diff
@@ -410,27 +410,44 @@ def get_equation_graph(self, refresh=False) -> nx.DiGraph:
 
-        # Create set of symbols for which we require equations
-        required_symbols = set()
+        def get_parents(node, graph=graph):
```
would this function not be useful as a global function ?

MauriceHendrix replied:
Perhaps I didn't  really look at the whole library. If you think it's useful feel free to put it elsewhere. This pull request was just me experimenting.

The main thing I'm wondering is whether it's correct. Are there ever any case where direct ancestors have an interdependence (e.g. x=5y, y=5)? (If so order is important if I'm able to generate C++ code)

jonc125 replied:
Not sure what you mean by direct ancestors having an interdependence?

MauriceHendrix replied:
ok so lets say I have `x = 5*y + 6*z`
ancestors:
y and z
 would it be possible for y and z then t have their own equations for example:
`y =2*z
z = 1`

jonc125 replied:
Almost certainly yes!

MauriceHendrix replied:
So we somehow need to make sure z ends up before y in the example somehow

MauriceHendrix replied:
Would a topological sort work?

jonc125 replied:
I thought get_equations_for was doing a topological sort? Using the nx graph library? Certainly that's the correct approach.

The trickier bit is determining a consistent relative ordering where there isn't a direct dependency.

MauriceHendrix replied:
It does but not with my flag. i've changed that now

MauriceHendrix commented:
I just added code to prevent unwanted duplicates. Had to also add test so put it in test_simple_model_units.py (may need to change) and pull in simple_model_units.cellml from the  investigate units branch.
Also split out test for duplicate input in test_hodgkin.py

jonc125 commented on cellmlmanip/model.py:
```diff
@@ -410,27 +410,46 @@ def get_equation_graph(self, refresh=False) -> nx.DiGraph:
         eqs = []
-        for symbol in sorted_symbols:
+
+        if sort_by_input_symbols:
```
If this is false, we're doing a lot of work above that's not needed!

MauriceHendrix replied:
Not really if it's false we still check the required_symbols, we're going to need those one way or another.
```        for symbol in to_search_in:
            # Ignore symbols we don't need
            if symbol not in required_symbols:
                continue```

jonc125 commented on cellmlmanip/model.py:
```diff
@@ -410,27 +410,46 @@ def get_equation_graph(self, refresh=False) -> nx.DiGraph:
-        required_symbols = set()
+        sorted_symbols = [s for s in nx.topological_sort(graph)]
+
+        def get_parents(node, graph=graph, sorted_symbols=sorted_symbols):
```
I'd like this at least to have a docstring, even if we don't make it a class method.

jonc125 commented on cellmlmanip/model.py:
```diff
@@ -410,27 +410,46 @@ def get_equation_graph(self, refresh=False) -> nx.DiGraph:
         Results are sorted in topographical order.
         :param symbols: the symbols to get the equations for
-        :param lexicographical_sort: indicates whether the result is sorted in lexicographical order first
+        :param sort_by_input_symbols: indicates whether the sorting in symbols determines the sorting of the output
```
Doing this manually just feels wrong. Surely there's something in the NetworkX library we can get to do this for us?

For instance, the [lexicographic](https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.lexicographical_topological_sort.html) sort takes a `key` argument to resolve ambiguities in the sort order. Would that work?

MauriceHendrix replied:
What you'd have to do is write a key function key such that 
given input [x1...xN] to the get_equations_for fuction
key(x1) < ... < ... key(xN) (which is straightforward) but also
key (y1) < ... < key(yN) where yn is a parent of xn

I don't really see any way of doing this without actually doing the sort unless I'm missing something?

jonc125 commented:
The purpose of this sorting is purely to make it easier to compare with pycml output right? So we're getting rid of it and reverting to lexicographical sort once we can generate new reference data? In which case it's probably not worth worrying about!

--
Written on a smartphone; please excuse brevity!

________________________________
From: MauriceHendrix <notifications@github.com>
Sent: Sunday, September 22, 2019 10:40:12 AM
To: ModellingWebLab/cellmlmanip <cellmlmanip@noreply.github.com>
Cc: Cooper, Jonathan <j.p.cooper@ucl.ac.uk>; Mention <mention@noreply.github.com>
Subject: Re: [ModellingWebLab/cellmlmanip] added input order preservation and exclusion list to get_equations_for (#85)


@MauriceHendrix commented on this pull request.

________________________________

In cellmlmanip/model.py<https://github.com/ModellingWebLab/cellmlmanip/pull/85#discussion_r326891968>:

>          """Get all equations for given collection of symbols

         Results are sorted in topographical order.
         :param symbols: the symbols to get the equations for
-        :param lexicographical_sort: indicates whether the result is sorted in lexicographical order first
+        :param sort_by_input_symbols: indicates whether the sorting in symbols determines the sorting of the output


What you'd have to do is write a key function key such that
given input [x1...xN] to the get_equations_for fuction
key(x1) < ... < ... key(xN) (which is straightforward) but also
key (y1) < ... < key(yN) where yn is a parent of xn

I don't really see any way of doing this without actually doing the sort unless I'm missing something?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ModellingWebLab/cellmlmanip/pull/85?email_source=notifications&email_token=AAPQJZ4Q3VNOQ7O27CFAWXDQK44PZA5CNFSM4IXGWHDKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFP6ULY#discussion_r326891968>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAPQJZ75E4HNB55IWSQNV6LQK44PZANCNFSM4IXGWHDA>.

MauriceHendrix commented:
> The purpose of this sorting is purely to make it easier to compare with pycml output right? So we're getting rid of it and reverting to lexicographical sort once we can generate new reference data? In which case it's probably not worth worrying about!

Yes I was going to contact you about that actually. in fact for now maybe don't even merge this in yet. Once I have the normal model properly working I'll write a test that can cope with equations not being in the same order and then test if we actually really need this or not.

jonc125 commented:
Sounds good!

--
Written on a smartphone; please excuse brevity!

________________________________
From: MauriceHendrix <notifications@github.com>
Sent: Sunday, September 22, 2019 10:55:43 AM
To: ModellingWebLab/cellmlmanip <cellmlmanip@noreply.github.com>
Cc: Cooper, Jonathan <j.p.cooper@ucl.ac.uk>; Mention <mention@noreply.github.com>
Subject: Re: [ModellingWebLab/cellmlmanip] added input order preservation and exclusion list to get_equations_for (#85)


The purpose of this sorting is purely to make it easier to compare with pycml output right? So we're getting rid of it and reverting to lexicographical sort once we can generate new reference data? In which case it's probably not worth worrying about!

Yes I was going to contact you about that actually. in fact for now maybe don't even merge this in yet. Once I have the normal model properly working I'll write a test that can cope with equations not being in the same order and then test if we actually really need this or not.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ModellingWebLab/cellmlmanip/pull/85?email_source=notifications&email_token=AAPQJZYXZCNFOGML4EMAW6DQK46J7A5CNFSM4IXGWHDKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7JCQYA#issuecomment-533866592>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAPQJZZECCMLQ7VQZIKUW5LQK46J7ANCNFSM4IXGWHDA>.

MauriceHendrix commented:
Right so it looks like this isn't necessary at all as it looks like I get things back in a suitable order (even if not the same as pycml)
